src/Arcus.Messaging.Pumps.ServiceBus/Extensions/IServiceCollectionExtensions.cs

```diff
@@ -728,5 +728,49 @@ public static class IServiceCollectionExtensions
 
             return services;
         }
+
+        /// <summary>
+        /// Adds a <see cref="IAzureServiceBusMessageHandler{TMessage}" /> implementation to process the messages from an <see cref="AzureServiceBusMessagePump"/> implementation.
+        /// resources.
+        /// </summary>
+        /// <typeparam name="TMessageHandler">The type of the implementation.</typeparam>
+        /// <typeparam name="TMessage">The type of the message that the message handler will process.</typeparam>
+        /// <param name="services">The collection of services to use in the application.</param>
+        /// <param name="messageContextFilter">The function that determines if the message handler should handle the message based on the context.</param>
+        public static IServiceCollection WithServiceBusMessageHandler<TMessageHandler, TMessage>(
+            this IServiceCollection services,
+            Func<AzureServiceBusMessageContext, bool> messageContextFilter)
+            where TMessageHandler : class, IAzureServiceBusMessageHandler<TMessage>
+            where TMessage : class
+        {
+            Guard.NotNull(services, nameof(services));
+            Guard.NotNull(messageContextFilter, nameof(messageContextFilter));
+
+            return services.WithMessageHandler<TMessageHandler, TMessage, AzureServiceBusMessageContext>(
+                messageContextFilter, serviceProvider => ActivatorUtilities.CreateInstance<TMessageHandler>(serviceProvider));
+        }
+
+        /// <summary>
+        /// Adds a <see cref="IAzureServiceBusMessageHandler{TMessage}" /> implementation to process the messages from an <see cref="AzureServiceBusMessagePump"/> implementation.
+        /// resources.
+        /// </summary>
+        /// <typeparam name="TMessageHandler">The type of the implementation.</typeparam>
+        /// <typeparam name="TMessage">The type of the message that the message handler will process.</typeparam>
+        /// <param name="services">The collection of services to use in the application.</param>
+        /// <param name="messageContextFilter">The function that determines if the message handler should handle the message based on the context.</param>
+        /// <param name="implementationFactory">The function that creates the message handler.</param>
+        public static IServiceCollection WithServiceBusMessageHandler<TMessageHandler, TMessage>(
```

Comment:
We should decouple this from Service Bus and only give them `MessageContext`

Reply:
Why is that? Isn't that the point of registering a service bus message handler? If people want a different context, shouldn't they use the `.WithMessageHandler` instead?

Reply:
This is what we want to avoid:
![image](https://user-images.githubusercontent.com/4345663/86323929-05d41200-bc3e-11ea-9267-ad6521586621.png)

Meaning, people need to have the opportunity to be able to use message handler with filtering that does not tightly-couple it to Service Bus. Do that mean we should remove the current on? No, but we need to have the filtering capability on `WithMessageHandler` at least.

That's why we decoupled it a while back, to make sure the same message handler can be used with another queuing tech.

Reply:
Yes, and the context filtering is available on the `.WithMessageHandler`, the `.WithServiceBusMessageHandler` only adds the specific service bus message context to the handler so we don't have to specify:

```csharp
.WithMessageHandler<OrdersMessageHandler, Order, ServiceBusMessageContext>()
```

Reply:
So, I'm not sure what I should remove/change here and if I implement something wrong here maybe?

Reply:
I've misread, didn't notice the other extensions sorry! No action to be taken then but I would at least add some tests using the non-SB approach since we're replacing those

Reply:
Yes, updated the existing test to verify more thoroughly the predication selection.